tests/client/test_retries.py

```diff
@@ -0,0 +1,191 @@
+        client.get("https://example.com/connect_error")
+
+
+def test_retries_enabled() -> None:
+    """
+    When retries are enabled, connection failures are retried on with
+    a fixed exponential backoff.
+    """
+    transport = MockTransport(num_failures=3)
+    client = httpx.Client(transport=transport, retries=3)
+    expected_elapsed_time = pytest.approx(0 + 0.5 + 1, rel=0.1)
+
+    response = client.get("https://example.com")
+    assert response.status_code == 200
+
+    response = client.get("https://example.com/connect_timeout")
+    assert response.status_code == 200
+    assert response.elapsed.total_seconds() == expected_elapsed_time
+
+    response = client.get("https://example.com/connect_error")
+    assert response.status_code == 200
+    assert response.elapsed.total_seconds() == expected_elapsed_time
+
+    with pytest.raises(httpx.ReadTimeout):
+        client.get("https://example.com/read_timeout")
+
+    with pytest.raises(httpx.NetworkError):
+        client.get("https://example.com/network_error")
+
+
+@pytest.mark.usefixtures("async_environment")
+async def test_retries_enabled_async() -> None:
+    # For test coverage purposes.
+    transport = AsyncMockTransport(num_failures=3)
+    client = httpx.AsyncClient(transport=transport, retries=3)
+    expected_elapsed_time = pytest.approx(0 + 0.5 + 1, rel=0.1)
+
+    # Connect exceptions are retried on with a backoff.
+    response = await client.get("https://example.com/connect_timeout")
+    assert response.status_code == 200
+    assert response.elapsed.total_seconds() == expected_elapsed_time
+
+    # Non-connect errors are not retried on.
+    with pytest.raises(httpx.ReadTimeout):
+        await client.get("https://example.com/read_timeout")
+
+
+def test_retries_exceeded() -> None:
+    """
+    When retries are enabled and connecting failures more than the configured number
+    of retries, connect exceptions are raised.
+    """
+    transport = MockTransport(num_failures=2)
+    client = httpx.Client(transport=transport, retries=1)
+
+    with pytest.raises(httpx.ConnectTimeout):
+        client.get("https://example.com/connect_timeout")
+
+    with pytest.raises(httpx.ConnectError):
+        client.get("https://example.com/connect_error")
+
+
+@pytest.mark.parametrize(
+    "method", ["HEAD", "GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS", "TRACE"]
+)
+def test_retries_methods(method: str) -> None:
+    """
+    Client retries on all HTTP methods.
+    """
+    transport = MockTransport(num_failures=1)
+    client = httpx.Client(transport=transport, retries=1)
+    response = client.request(method, "https://example.com/connect_timeout")
+    assert response.status_code == 200
+
+
+@pytest.mark.parametrize(
+    "factor, expected",
+    [
+        (0.1, [0, 0.1, 0.2, 0.4, 0.8]),
+        (0.2, [0, 0.2, 0.4, 0.8, 1.6]),
+        (0.5, [0, 0.5, 1.0, 2.0, 4.0]),
+    ],
+)
+def test_exponential_backoff(factor: float, expected: List[int]) -> None:
```

Comment:
Since this is the only test case that requires a private import, perhaps we could put this one in `tests/test_utils.py`?

Working towards making sure we're only using `import httpx` in our test cases, but the utils cases are an exception to that.